cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SystemProgramManagementService.java

```diff
@@ -0,0 +1,197 @@
+
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.google.inject.Inject;
+import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.store.RunRecordDetail;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.ProgramId;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * System program management service for ensuring services are running/stopped as expected
+ */
+public class SystemProgramManagementService extends AbstractScheduledService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemProgramManagementService.class);
+
+  private final long scheduleInterval;
+  private final CConfiguration cConf;
+  private final ProgramRuntimeService programRuntimeService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final AtomicReference<Map<ProgramId, Map<String, String>>> programsEnabled;
+
+  @Inject
+  SystemProgramManagementService(CConfiguration cConf, ProgramRuntimeService programRuntimeService,
+                                 ProgramLifecycleService programLifecycleService) {
+    this.cConf = cConf;
+    this.scheduleInterval = cConf.getLong(Constants.AppFabric.PROGRAM_SCAN_INTERVAL);
+    this.programRuntimeService = programRuntimeService;
+    this.programLifecycleService = programLifecycleService;
+    this.programsEnabled = new AtomicReference<>();
+  }
+
+  /**
+   * Sets the map of services that are currently enabled along with their runtime args.
+   * The services that are not present in map will be stopped during the next run of the service.
+   *
+   * @param programsEnabled
+   */
+  public void setProgramsEnabled(Map<ProgramId, Map<String, String>> programsEnabled) {
+    this.programsEnabled.set(new HashMap<>(programsEnabled));
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    if (programsEnabled.get() == null) {
+      LOG.debug("Services to run not yet set");
+      return;
+    }
+    try {
+      reconcileServices();
+    } catch (Exception ex) {
+      //catch here so a single iteration doesn't stop the service
+      LOG.error("Error in reconciling services", ex);
+    }
+  }
+
+  private void reconcileServices() throws Exception {
+    //take a copy
+    Map<ProgramId, Map<String, String>> enabledServicesMap = new HashMap<>(this.programsEnabled.get());
+    Set<ProgramId> enabledServices = enabledServicesMap.keySet();
+    Set<ProgramId> servicesToStop = new HashSet<>();
```

Comment:
You can just keep a `Set<ProgramRunId>` for programRunsToStop.

Reply:
Done